samples/bluetooth/iso_broadcast/src/main.c

```diff
@@ -197,22 +215,116 @@ int main(void)
+				/* Use sequence number 0, to synchronize current stream with the
+				 * current payload counter in the ISO group of streams.
+				 */
+				is_first_iso_data_send = true;
+				seq_num[chan] = 0U;
+			}
+
+			/* Send ISO data with timestamp, use sequence number 0 for first ISO data on
+			 * any stream.
+			 */
+			if (!is_first_iso_data_send) {
+				ret = bt_iso_chan_send_ts(&bis_iso_chan[chan], buf, seq_num[chan],
+							  ts);
+			} else {
+				ret = bt_iso_chan_send(&bis_iso_chan[chan], buf, 0U);
+			}
+
 			if (ret < 0) {
 				printk("Unable to broadcast data on channel %u"
 				       " : %d", chan, ret);
 				net_buf_unref(buf);
 				return 0;
 			}
 
+			/* Get Tx Sync, to verify that all streams have their ISO data at the same
+			 * timestamp.
+			 */
+			ret = bt_iso_chan_get_tx_sync(&bis_iso_chan[chan], &tx_info);
+			if (ret) {
+				printk("Failed to get tx sync after send (%d)\n", ret);
+				return 0;
+			}
+
+			/* Timestamp delta with the Controller */
+			if (!delta) {
+				delta = tx_info.ts - ts;
+			} else {
+				int64_t drift = ((delta - tx_info.ts + ts) / BIG_SDU_INTERVAL_US) /
+						1000000U;
+
+				/* NOTE: As the `ts` is just a increase in software in this sample,
+				 *       we will not see any drift.
+				 */
+				if (drift) {
+					printk("Drift: %d ppm.\n", drift);
+				}
+
+				/* NOTE: Change in delta (drift) for a duration of SDU intervals can
+				 *       be used to synchronize the clock with the Controller.
+				 */
+				static uint8_t countdown = 100U;
+				static int64_t drift_us;
+
+				drift_us += delta - tx_info.ts + ts;
+
+				if (!drift_comp_us && !countdown--) {
+					countdown = 100U;
+					drift_comp_us = drift_us / (countdown + 1U);
+					drift_us = 0U;
+
+					printk("Drift Compensate %d us\n", drift_comp_us);
+				}
+			}
+
+			printk("Tx chan %u: seq num %u -> %u ts %u -> %u offset %u\n", chan,
+			       seq_num[chan], tx_info.seq_num, ts, tx_info.ts, tx_info.offset);
+
+			/* Permit a reasonable clock jitter (at least Bluetooth active clock
+			 * jitter), in the verification of timestamp.
+			 */
+			uint32_t jitter_us = BIG_SDU_INTERVAL_US;
+
+			/* Validate ISO data being enqueued at the required timestamp.
+			 * Ignore checking first ISO data sent.
+			 */
+			if (!is_first_iso_data_send &&
+			    !IN_RANGE(tx_info.ts, (ts + delta - jitter_us),
+				      (ts + delta + jitter_us))) {
+				printk("Mismatch in tx sync timestamp! (%u != %u)\n", (ts + delta),
+				       tx_info.ts);
+				return 0;
+			}
+
+			seq_num[chan]++;
 		}
 
 		if ((iso_send_count % CONFIG_ISO_PRINT_INTERVAL) == 0) {
 			printk("Sending value %u\n", iso_send_count);
 		}
 
 		iso_send_count++;
-		seq_num++;
+
+		uint32_t sleep_us = BIG_SDU_INTERVAL_US - (drift_comp_us * 2);
+
+		drift_comp_us = 0U;
+
+		uint32_t ts_prev = k_ticks_to_us_floor32(sys_clock_tick_get());
+
+		k_sleep(K_USEC(sleep_us));
+
+		uint32_t ts_curr = k_ticks_to_us_floor32(sys_clock_tick_get());
+
+		ts += MAX(sleep_us, (ts_curr - ts_prev));
```

Comment:
Why are we adding sleep code the sample? When using timestamps there is no need to sleep, the controller would know when to send the payloads based upon the received timestamp

Reply:
The code here is simulating the relative drift in the sampling clock with respect to the Controller reference clock. And expects the Controller to handle the below specification statement:

```
When an HCI ISO Data packet sent by the Host does not contain a Time_Stamp or
the Time_Stamp value is not based on the Controller's clock, the Controller should
determine the CIS or BIS event to be used to transmit the SDU contained in that packet
based on the time of arrival of that packet.
```

The `k_sleep` is used to simulate the time of arrival of ISO data packets to the Controller.

Reply:
But if we are using timestamps, then time of arrival is likely not used anyways. Personally I would therefore remove this to reduce the complexity of the sample